### PR TITLE
Make NewTextPainter actually work

### PR DESCRIPTION
**Pull request by abarth**
The NewTextPainter is still disabled by default. A future patch will flip the
flag to enable it.

This patch uses a new approach to writing bindings by encoding data into array
buffers. This approach is more efficient than the existing IDL based approach.
If this works out well, we should convert our other performance-sensitive
interfaces to this approach in future patches.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

**Comment by abarth** [comment]
As far as I can tell, this fully works, but I'll flip the flag separately.

File: sky/engine/core/dart/text.dart [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  dotted,
+
+  /// Draw a dashed line
+  dashed,
+
+  /// Draw a sinusoidal line
+  wavy
+}
+
+// This encoding must match ParagraphBuilder::pushStyle.
```
Comment (Hixie) [inline sky/engine/core/dart/text.dart]: I assume you don't mean the ParagraphBuilder::pushStyle in this file, but the C++ version? Clarify the comment.

File: sky/engine/core/dart/text.dart [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+Int32List _encodeTextStyle(Color color,
+                           List<TextDecoration> decoration,
+                           Color decorationColor,
+                           TextDecorationStyle decorationStyle,
+                           FontWeight fontWeight,
+                           FontStyle fontStyle,
+                           String fontFamily,
+                           double fontSize) {
+  Int32List result = new Int32List(7);
```
Comment (Hixie) [inline sky/engine/core/dart/text.dart]: Document this protocol somewhere.

File: sky/engine/core/dart/text.dart (hunk 1 above)
Comment (abarth) [inline sky/engine/core/dart/text.dart]: That's correct.

File: sky/engine/core/text/Paragraph.cpp [hunk 3]
```diff
@@ -60,11 +60,18 @@ void Paragraph::layout()
 void Paragraph::paint(Canvas* canvas, const Offset& offset)
 {
     // Very simplified painting to allow painting an arbitrary (layer-less) subtree.
+    RenderBox* box = firstChildBox();
     GraphicsContext context(canvas->skCanvas());
+
+    // Very simplified painting to allow painting an arbitrary (layer-less) subtree.
```
Comment (Hixie) [inline sky/engine/core/text/Paragraph.cpp]: This comment now appears twice?

File: sky/engine/core/text/ParagraphBuilder.cpp [hunk 4]
```diff
@@ -41,43 +23,168 @@ RenderParagraph* createRenderParagraph(RenderStyle* parentStyle)
...
+    int32_t mask = encoded[0];
+
+    if (mask) {
+      RefPtr<RenderStyle> style = RenderStyle::clone(m_renderParagraph->style());
+
+      if (mask & kTextAlignMask)
+        style->setTextAlign(static_cast<ETextAlign>(encoded[kTextAlignIndex]));
+
+      if (mask & kTextBaselineMask) {
+        // TODO(abarth): Implement TextBaseline. It's not clear whether anything
```
Comment (Hixie) [inline sky/engine/core/text/ParagraphBuilder.cpp]: it wasn't.

**Comment by Hixie** [comment]
LGTM, though I didn't study it very carefully.

File: sky/engine/core/dart/text.dart (hunk 2 above)
Comment (abarth) [inline sky/engine/core/dart/text.dart]: Done.

File: sky/engine/core/text/Paragraph.cpp (hunk 3 above)
Comment (abarth) [inline sky/engine/core/text/Paragraph.cpp]: Fixed.

**Pull request merged**